src/sim/TrackInitializerStore.cc

```diff
@@ -0,0 +1,225 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2020 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file TrackInitializerStore.cc
+//---------------------------------------------------------------------------//
+#include "TrackInitializerStore.hh"
+
+#include <numeric>
+#include "detail/InitializeTracks.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with the number of tracks and the maximum number of elements to
+ * allocate on device.
+ */
+TrackInitializerStore::TrackInitializerStore(size_type        num_tracks,
+                                             size_type        capacity,
+                                             constVecPrimary& primaries)
+    : initializers_(capacity)
+    , parent_(capacity)
+    , vacancies_(num_tracks)
+    , secondary_counts_(num_tracks)
+    , primaries_(primaries)
+    , num_primaries_(primaries.size())
+{
+    // Start with an empty vector of track initializers and parent thread IDs
+    initializers_.resize(0);
+    parent_.resize(0);
+
+    // Initialize vacancies to mark all track slots as initially empty
+    std::vector<size_type> host_vacancies(vacancies_.size());
+    std::iota(host_vacancies.begin(), host_vacancies.end(), 0);
+    vacancies_.copy_to_device(make_span(host_vacancies));
+
+    // Initialize the track counter for each event as the number of primary
+    // particles in that event
+    std::vector<ull_int> host_track_counter;
+    for (const auto& primary : primaries_)
+    {
+        const auto event_id = primary.event_id.get();
+        if (host_track_counter.size() <= event_id)
+        {
+            host_track_counter.resize(event_id + 1);
+        }
+        ++host_track_counter[event_id];
+    }
+    track_counter_ = DeviceVector<ull_int>(host_track_counter.size());
+    track_counter_.copy_to_device(make_span(host_track_counter));
```

Comment:
This is an excellent "temporary" workaround to the event-specific thread ID that we'll put in a new MR for.